File: leetcode/medium/17. Letter Combinations of a Phone Number/solution.cc
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits == "") return {};
        vector<string> res = {""};
        vector<string> table = { "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };

        for (int i = 0; i < digits.size(); i++) {
            vector<string> temp;
            string s = table[digits[i] - '0'];
            
            for (char ch: s) {
                for (int j = 0; j < res.size(); j++) {
                    temp.push_back(res[j] + ch);
                }
            }
            swap(temp, res);
        }
        return res;
    }
};

File: leetcode/medium/17. Letter Combinations of a Phone Number/solution.js
/** https://leetcode.com/problems/letter-combinations-of-a-phone-number/
 * @param {string} digits
 * @return {string[]}
 */
function letterCombinations(digits) {
    const table = {
        2: 'abc',
        3: 'def',
        4: 'ghi',
        5: 'jkl',
        6: 'mno',
        7: 'pqrs',
        8: 'tuv',
        9: 'wxyz',
    };

    let ret = table[digits[0]];
    if (!ret) return [];
    ret = ret.split('');

    for (let i = 1; i < digits.length; i += 1) {
        ret = stringAdder(ret, table[digits[i]].split(''));
    }

    return ret;
};

function stringAdder(arrayA, arrayB) {
    return arrayA.flatMap(valA => arrayB.map(valB => valA + valB));
}